### PR TITLE
[bitnami/cassandra] Additional Storage Class and Commit Log Directory

### PR DESCRIPTION
**Pull request by mehmetsalgar**
**Description of the change**
It is recommended in Cassandra documentation to seperate the storage for data (SSTable, etc) and the Commit Log Sorage.
[Commit Log Storage](https://cassandra.apache.org/doc/latest/architecture/storage_engine.html)
```
* NOTE: In the event of an unexpected shutdown, Cassandra can lose up to the sync period or more if the sync is delayed. If using “batch” mode, it is recommended to store commitlogs in a separate, dedicated device.
```
So I like to configure another StorageClass and and Volume/VolumeMount for Commit Log Storage but unfortunately volumeClaimTemplates in statefulset.yaml is not design that way. Is it possible to changes it so it will be possible.

_helpers.tpl
stateful.yaml
values.yaml

are changed so the template can accept additional Storage Class for Commit Logs and configure necessary Environment Parameters

**Benefits**

It is documented in Cassandra Documentation writing Data and Commits Logs should not compete for IO and use seperate Drives, which is achieved in different StorageClasses in Cloud

**Possible drawbacks**

Not known for me.

**Applicable issues**
#5637 

**Additional information**

**Checklist** 
- [X ] Title of the PR starts with chart name (e.g. `[bitnami/chart]`)

### CODE REVIEW

File: bitnami/cassandra/templates/statefulset.yaml [hunk 1]
```diff
@@ -103,10 +123,14 @@ spec:
           volumeMounts:
             - name: data
               mountPath: {{ .Values.persistence.mountPath }}
+            {{- if .Values.persistence.commitStorageClass }}
+            - name: commitlog
+              mountPath: {{ .Values.persistence.commitLogMountPath }}
+        {{- end }}
```
Comment (Mauraza) [inline bitnami/cassandra/templates/statefulset.yaml:129]: Could you put this line and the same level as the 'if'?

File: bitnami/cassandra/templates/statefulset.yaml [hunk 2]
```diff
@@ -103,10 +123,14 @@ spec:
...
         {{- end }}
         {{- if .Values.initContainers }}
         {{- include "common.tplvalues.render" (dict "value" .Values.initContainers "context" $) | nindent 8 }}
-        {{- end }}
+      {{- end }}
```
Comment (Mauraza) [inline bitnami/cassandra/templates/statefulset.yaml:133]: add tabulation

File: bitnami/cassandra/templates/statefulset.yaml [hunk 3]
```diff
@@ -269,35 +297,39 @@ spec:
...
-              {{- end }}
+            {{- end }}
             {{- if .Values.cluster.enableRPC }}
             - name: thrift
               containerPort: {{ .Values.containerPorts.thrift | default "9160" }}
               {{- if .Values.hostNetwork }}
               hostPort: {{ .Values.containerPorts.thrift }}
-              {{- end }}
             {{- end }}
+          {{- end }}
```
Comment (Mauraza) [inline bitnami/cassandra/templates/statefulset.yaml:322]: Could you revert these changes? they are only tabulations

File: bitnami/cassandra/values.yaml [hunk 4]
```diff
@@ -17,7 +17,7 @@ image:
   ## Bitnami Cassandra image tag
   ## ref: https://github.com/bitnami/bitnami-docker-cassandra#supported-tags-and-respective-dockerfile-links
   ##
-  tag: 3.11.10-debian-10-r23
+  tag: 3.11.10-debian-10-r6
```
Comment (Mauraza) [inline bitnami/cassandra/values.yaml:20]: Could you change the tag to `3.11.10-debian-10-r23`?

File: bitnami/cassandra/values.yaml [hunk 5]
```diff
@@ -142,8 +156,8 @@ volumePermissions:
   enabled: false
   image:
     registry: docker.io
-    repository: bitnami/bitnami-shell
-    tag: "10"
+    repository: bitnami/minideb
+    tag: buster
```
Comment (Mauraza) [inline bitnami/cassandra/values.yaml:160]: could you revert these changes?

File: bitnami/cassandra/templates/statefulset.yaml [hunk 6]
```diff
@@ -395,14 +427,31 @@ spec:
...
+        accessModes:
+          {{- range .Values.persistence.accessModes }}
+          - {{ . | quote }}
+        {{- end }}
+        resources:
+          requests:
+            storage: {{ .Values.persistence.commitLogsize | quote }}
+      {{- include "cassandra.commitstorage.class" (dict "persistence" .Values.persistence "global" .Values.global) | nindent 8 }}
   {{- end }}
+  {{- end }}
```
Comment (Mauraza) [inline bitnami/cassandra/templates/statefulset.yaml:457]: could you add a new line?

**Review by Mauraza — COMMENTED** [review]
Hi @mehmetsalgar,

Thanks a lot for the contributions! I added some comments about tabulations and the tags of the image. Could you check that?

File: bitnami/cassandra/Chart.yaml [hunk 7]
```diff
@@ -23,4 +23,4 @@ name: cassandra
 sources:
   - https://github.com/bitnami/bitnami-docker-cassandra
   - http://cassandra.apache.org
-version: 7.3.3
+version: 7.3.4
```
Comment (Mauraza) [inline bitnami/cassandra/Chart.yaml:26]: this change adds a new feature could you increase the version to 7.4.0?

**Review by Mauraza — COMMENTED** [review]
sorry, I forgot that you also have to upgrade the version to 7.4.0

File: bitnami/cassandra/templates/statefulset.yaml [hunk 8]
```diff
@@ -4,9 +4,9 @@ metadata:
   name: {{ include "common.names.fullname" . }}
   namespace: {{ .Release.Namespace }}
   labels: {{- include "common.labels.standard" . | nindent 4 }}
-    {{- if .Values.commonLabels }}
-    {{- include "common.tplvalues.render" ( dict "value" .Values.commonLabels "context" $ ) | nindent 4 }}
-    {{- end }}
+  {{- if .Values.commonLabels }}
+  {{- include "common.tplvalues.render" ( dict "value" .Values.commonLabels "context" $ ) | nindent 4 }}
+  {{- end }}
```
[the remark this answers is not in the document]
Comment (mehmetsalgar) [reply bitnami/cassandra/templates/statefulset.yaml:9]: Interesting, I actually didn't change anything here, for some reason indention is changed and GiT Annotations say it is original code state from 'Toledano', may be an IntelliJ magic but I will try to revert

File: bitnami/cassandra/templates/statefulset.yaml [hunk 9]
```diff
@@ -23,7 +23,7 @@ spec:
     {{- else if .Values.rollingUpdatePartition }}
     rollingUpdate:
       partition: {{ .Values.rollingUpdatePartition }}
-    {{- end }}
+  {{- end }}
```
[the remark this answers is not in the document]
Comment (mehmetsalgar) [reply bitnami/cassandra/templates/statefulset.yaml]: Done

File: bitnami/cassandra/templates/statefulset.yaml [hunk 10]
```diff
@@ -32,13 +32,13 @@ spec:
...
-        {{- if .Values.metrics.podAnnotations }}
-        {{- toYaml .Values.metrics.podAnnotations | nindent 8 }}
-        {{- end }}
+      {{- if .Values.podAnnotations }}
+      {{- toYaml .Values.podAnnotations | nindent 8 }}
+      {{- end }}
+      {{- if .Values.metrics.podAnnotations }}
+      {{- toYaml .Values.metrics.podAnnotations | nindent 8 }}
       {{- end }}
+    {{- end }}
```
[the remark this answers is not in the document]
Comment (mehmetsalgar) [reply bitnami/cassandra/templates/statefulset.yaml]: Done

File: bitnami/cassandra/templates/statefulset.yaml (hunk 1 above)
Comment (mehmetsalgar) [reply bitnami/cassandra/templates/statefulset.yaml]: Done

File: bitnami/cassandra/templates/statefulset.yaml (hunk 2 above)
Comment (mehmetsalgar) [reply bitnami/cassandra/templates/statefulset.yaml]: Done

File: bitnami/cassandra/templates/statefulset.yaml (hunk 3 above)
Comment (mehmetsalgar) [reply bitnami/cassandra/templates/statefulset.yaml]: Done

File: bitnami/cassandra/values.yaml [hunk 11]
```diff
@@ -475,7 +489,7 @@ metrics:
     registry: docker.io
     pullPolicy: IfNotPresent
     repository: bitnami/cassandra-exporter
-    tag: 2.3.4-debian-10-r331
+    tag: 2.3.4-debian-10-r315
```
[the remark this answers is not in the document]
Comment (mehmetsalgar) [reply bitnami/cassandra/values.yaml]: Done

File: bitnami/cassandra/templates/statefulset.yaml (hunk 6 above)
Comment (mehmetsalgar) [reply bitnami/cassandra/templates/statefulset.yaml]: Done

File: bitnami/cassandra/values.yaml (hunk 5 above)
Comment (mehmetsalgar) [reply bitnami/cassandra/values.yaml]: Done

File: bitnami/cassandra/values.yaml (hunk 4 above)
Comment (mehmetsalgar) [reply bitnami/cassandra/values.yaml]: Done

File: bitnami/cassandra/templates/statefulset.yaml [hunk 12]
```diff
@@ -130,7 +154,7 @@ spec:
                   export CASSANDRA_IGNORE_INITDB_SCRIPTS=1
               fi
               /opt/bitnami/scripts/cassandra/entrypoint.sh /opt/bitnami/scripts/cassandra/run.sh
-            {{- end }}
+              {{- end }}
```
[the remark this answers is not in the document]
Comment (mehmetsalgar) [reply bitnami/cassandra/templates/statefulset.yaml]: Done

File: bitnami/cassandra/templates/statefulset.yaml [hunk 13]
```diff
@@ -211,9 +235,13 @@ spec:
...
               value: {{ .Values.containerPorts.cql | quote }}
             {{- end }}
+            {{- if .Values.persistence.commitStorageClass }}
+            - name: CASSANDRA_COMMITLOG_DIR
+              value: {{ .Values.persistence.commitLogMountPath | quote }}
+            {{- end }}
             {{- if .Values.extraEnvVars }}
             {{- include "common.tplvalues.render" (dict "value" .Values.extraEnvVars "context" $) | nindent 12 }}
-            {{- end }}
+          {{- end }}
```
[the remark this answers is not in the document]
Comment (mehmetsalgar) [reply bitnami/cassandra/templates/statefulset.yaml]: Done

File: bitnami/cassandra/templates/statefulset.yaml [hunk 14]
```diff
@@ -222,7 +250,7 @@ spec:
             {{- if .Values.extraEnvVarsSecret }}
             - secretRef:
                 name: {{ include "common.tplvalues.render" (dict "value" .Values.extraEnvVarsSecret "context" $) }}
-            {{- end }}
+          {{- end }}
```
[the remark this answers is not in the document]
Comment (mehmetsalgar) [reply bitnami/cassandra/templates/statefulset.yaml]: Done

File: bitnami/cassandra/Chart.yaml (hunk 7 above)
Comment (mehmetsalgar) [reply bitnami/cassandra/Chart.yaml]: Done

**Comment by Mauraza** [comment]
Hi @mehmetsalgar,

I was testing your solutions, and I have a question, in your solutions is it necessary to use a `PVC Storage Class for Cassandra Commit Log volume`  could be possible to use this with the dynamic provisioning (without the commitStorageClass)? 
Thanks!

**Comment by mehmetsalgar** [comment]
Sure, if the commitStorageClass has as value "" or "-", it will use dynamic provisioning. Which in production environment, will not make too much sense while the idea of the solution is for Cassandra Data directory and Commit Logs directory uses different physical drives. If both goes to same default provisioning, they could get the same physical device, which is ok for local testing but not so good for Production.

**Comment by Mauraza** [comment]
Hi @mehmetsalgar,

I'm trying to test your solutions with this command 
```
helm install cassandra cassandra/ --set persistence.commitStorageClass='' --set persistence.commitLogsize='2Gi' --set persistence.commitLogMountPath='/bitnami/cassandra/commitlog'
```

But the `Commit Logs directory` never appears, could you tell me what am I doing wrong?

**Comment by mehmetsalgar** [comment]
It seems the check

{{- if .Values.persistence.commitStorageClass }}

accepts 

--set persistence.commitStorageClass=''

as none existent value....

to use the default provisionier, you have to use 

--set persistence.commitStorageClass='-'

File: bitnami/cassandra/templates/_helpers.tpl [hunk 15]
```diff
@@ -82,3 +82,25 @@ cassandra: cluster.seedCount
...
+{{- end -}}
+
+{{- if $storageClass -}}
+  {{- if (eq "-" $storageClass) -}}
+      {{- printf "storageClassName: \"\"" -}}
+  {{- else }}
+      {{- printf "storageClassName: %s" $storageClass -}}
+  {{- end -}}
+{{- end -}}
+{{- end -}}
```
Comment (Mauraza) [inline bitnami/cassandra/templates/_helpers.tpl:106]: add a new line at the end of the file.

File: bitnami/cassandra/templates/statefulset.yaml [hunk 16]
```diff
@@ -103,6 +123,10 @@ spec:
           volumeMounts:
             - name: data
               mountPath: {{ .Values.persistence.mountPath }}
+            {{- if .Values.persistence.commitStorageClass }}
```
Comment (Mauraza) [inline bitnami/cassandra/templates/statefulset.yaml:126]: as the `storageClass` is not mandatory, it is better to use them `commitLogMountPath` because otherwise if you run:
```
helm template cassandra cassandra/ --set persistence.commitStorageClass='-' --set persistence.commitLogsize='2Gi' --set persistence.commitLogMountPath='/bitnami/cassandra/commitlog'
```
at the end of the file appear this line
```
 storageClassName: ""
```
And the pvc `commitlog-cassandra-0` never get bound

Could you change this for `{{- if .Values.persistence.commitLogMountPath }}`?

File: bitnami/cassandra/templates/statefulset.yaml [hunk 17]
```diff
@@ -211,6 +235,10 @@ spec:
             - name: CASSANDRA_CQL_PORT_NUMBER
               value: {{ .Values.containerPorts.cql | quote }}
             {{- end }}
+            {{- if .Values.persistence.commitStorageClass }}
```
Comment (Mauraza) [inline bitnami/cassandra/templates/statefulset.yaml:238]: Could you change this for `{{- if .Values.persistence.commitLogMountPath }}`?

File: bitnami/cassandra/templates/statefulset.yaml [hunk 18]
```diff
@@ -298,6 +326,10 @@ spec:
           volumeMounts:
             - name: data
               mountPath: {{ .Values.persistence.mountPath }}
+            {{- if .Values.persistence.commitStorageClass }}
```
Comment (Mauraza) [inline bitnami/cassandra/templates/statefulset.yaml:329]: Could you change this for `{{- if .Values.persistence.commitLogMountPath }}`?

File: bitnami/cassandra/templates/statefulset.yaml [hunk 19]
```diff
@@ -404,5 +436,22 @@ spec:
         resources:
           requests:
             storage: {{ .Values.persistence.size | quote }}
-        {{- include "common.storage.class" (dict "persistence" .Values.persistence "global" .Values.global) | nindent 8 }}
+    {{- include "common.storage.class" (dict "persistence" .Values.persistence "global" .Values.global) | nindent 8 }}
+    {{- if .Values.persistence.commitStorageClass }}
```
Comment (Mauraza) [inline bitnami/cassandra/templates/statefulset.yaml:440]: Could you change this for `{{- if .Values.persistence.commitLogMountPath }}`?

**Review by Mauraza — COMMENTED** [review]
Hi @mehmetsalgar, 
Could you check the comments?

**Comment by mehmetsalgar** [comment]
Done

**Comment by Mauraza** [comment]
Hi @mehmetsalgar,

Thanks again for the PR. LGTM!! :+1:
